snapshots/stream.go

```diff
@@ -0,0 +1,107 @@
+package snapshots
+
+import (
+	"bufio"
+	"compress/zlib"
+	"io"
+
+	protoio "github.com/gogo/protobuf/io"
+	"github.com/gogo/protobuf/proto"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+const (
+	// Do not change chunk size without new snapshot format (must be uniform across nodes)
+	snapshotChunkSize  = uint64(10e6)
+	snapshotBufferSize = int(snapshotChunkSize)
+)
+
+// StreamWriter set up a stream pipeline to serialize snapshot nodes:
+// ExportNode -> delimited Protobuf -> zlib -> buffer -> chunkWriter -> chan io.ReadCloser
+type StreamWriter struct {
+	chunkWriter *ChunkWriter
+	bufWriter   *bufio.Writer
+	zWriter     *zlib.Writer
+	protoWriter protoio.WriteCloser
+}
+
+// NewStreamWriter set up a stream pipeline to serialize snapshot nodes.
+func NewStreamWriter(ch chan<- io.ReadCloser) *StreamWriter {
+	chunkWriter := NewChunkWriter(ch, snapshotChunkSize)
+	bufWriter := bufio.NewWriterSize(chunkWriter, snapshotBufferSize)
+	zWriter, err := zlib.NewWriterLevel(bufWriter, 7)
```

Comment:
why 7? Should this be configurable?

Reply:
- `7` is the level picked previously, and to keep backward compatibility we better not change it I think.
- this number should be part of the consensus between the nodes because each node should create an identical snapshot for the same height, so shouldn't be a configurable per node.
- I'll make it a constant and add some comments.